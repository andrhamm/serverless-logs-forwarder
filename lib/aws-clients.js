/* eslint-disable global-require */
// export { default as AWSXRay } from 'aws-xray-sdk';
// exports.AWSXRay.captureHTTPsGlobal(require('http'));
// exports.AWSXRay.captureHTTPsGlobal(require('https'));

import AWS from 'aws-sdk';
// const AWS = exports.AWSXRay.captureAWS(require('aws-sdk'));

export const cloudformation = new AWS.CloudFormation({ apiVersion: '2010-05-15', region: 'us-east-1' });
export const cloudWatchLogs = new AWS.CloudWatchLogs({ apiVersion: '2014-03-28' });
export const kms = new AWS.KMS({ apiVersion: '2014-11-01' });
export const s3 = new AWS.S3({ apiVersion: '2006-03-01' });
// export const sqs = new AWS.SQS({ apiVersion: '2012-11-05' });
export const sns = new AWS.SNS({ apiVersion: '2010-03-31' });
// export const http = exports.AWSXRay.captureHTTPs(require('http'));
// export const https = exports.AWSXRay.captureHTTPs(require('https'));
