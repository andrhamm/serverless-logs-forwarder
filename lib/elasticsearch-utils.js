import elasticsearch from 'elasticsearch';

export const client = opts => new elasticsearch.Client({
  // log: 'trace', // uncomment for debug logging
  apiVersion: '5.6',
  protocol: 'http',
  requestTimeout: 6000,
  ...opts,
});
