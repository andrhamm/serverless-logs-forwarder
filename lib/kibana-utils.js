import axios from 'axios';

const getDefaultIndexSettings = async (kibanaApiBase) => {
  const url = `${kibanaApiBase}/index-pattern/cloudwatchlogs-*`;

  const req = {
    method: 'get',
    url,
    headers: {
      'kbn-version': '5.0.1',
      'kbn-xsrf': 'cloudWatchlogs-*',
    },
  };

  const resp = await axios(req);

  // eslint-disable-next-line no-underscore-dangle
  return resp.data._source;
};

export const saveIndexPattern = async ({ stackName, stage, kibanaApiBase }) => {
  const defaultIndexSettings = await getDefaultIndexSettings(kibanaApiBase);

  const servicePrefix = new RegExp(`^(.*?)(?:-${stage})?$`, 'i').exec(stackName)[1];

  const indexPattern = `cloudwatchlogs-${servicePrefix}-*`;

  const url = `${kibanaApiBase}/index-pattern/${indexPattern}`;

  const req = {
    method: 'put',
    url,
    headers: {
      'kbn-version': '5.0.1',
      'kbn-xsrf': indexPattern,
    },
    data: {
      ...defaultIndexSettings,
      title: indexPattern,
    },
  };

  return axios(req);
};
