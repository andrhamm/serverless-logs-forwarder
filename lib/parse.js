import zlib from 'zlib';
import { tryParseJson } from './common';

export const parsePayload = (record) => {
  // return Buffer.from(record.kinesis.data, 'base64').toString('utf8')
  const payload = Buffer.from(record.kinesis.data, 'base64');
  const json = (zlib.gunzipSync(payload)).toString('utf8');
  return JSON.parse(json);
};

// filter out CONTROL_MESSAGE records
export const extractRecords = records => records
  .map(exports.parsePayload).filter(r => r.messageType === 'DATA_MESSAGE');

// logGroup looks like this:
//    "logGroup": "/aws/lambda/service-env-funcName"
export const parseFunctionName = logGroup => logGroup.split('/').reverse()[0];

// logStream looks like this:
//    "logStream": "2016/08/17/[76]afe5c000d5344c33b5d88be7a4c55816"
export const parseLambdaVersion = (logStream) => {
  const start = logStream.indexOf('[');
  const end = logStream.indexOf(']');
  return logStream.substring(start + 1, end);
};

/* eslint-disable max-len */
// a Lambda function log message looks like this:
//    "2017-04-26T10:41:09.023Z db95c6da-2a6c-11e7-9550-c91b65931beb\tloading index.html...\n"
// but there are START, END and REPORT messages too:
//    "START RequestId: 67c005bb-641f-11e6-b35d-6b6c651a2f01 Version: 31\n"
//    "END RequestId: 5e665f81-641f-11e6-ab0f-b1affae60d28\n"
//    "REPORT RequestId: 5e665f81-641f-11e6-ab0f-b1affae60d28\tDuration: 1095.52 ms\tBilled Duration: 1100 ms \tMemory Size: 128 MB\tMax Memory Used: 32 MB\t\n"
/* eslint-enable max-len */
export const parseLogMessage = (logEvent) => {
  const { extractedFields } = logEvent;

  if (!logEvent.id ||
      extractedFields.event.startsWith('START RequestId') ||
      extractedFields.event.startsWith('END RequestId') ||
      extractedFields.event.startsWith('REPORT RequestId')) {
    return null;
  }

  // eslint-disable-next-line prefer-const
  let { event, timestamp } = extractedFields;

  const requestId = extractedFields.request_id;

  delete extractedFields.request_id;
  delete extractedFields.event;
  delete extractedFields.timestamp;

  const fields = extractedFields;

  const jsonFields = tryParseJson(event);
  if (jsonFields) {
    if (typeof jsonFields === 'object') {
      if (jsonFields.message) {
        event = jsonFields.message;
        delete jsonFields.message;
      } else {
        event = null;
      }
      Object.assign(fields, jsonFields);
    }
  }

  const level = (fields.level || 'info').toLowerCase();

  delete fields.level;


  const result = {
    level,
    message: event,
    request_id: requestId,
    '@timestamp': new Date(timestamp),
  };

  if (Object.keys(fields).length > 0) {
    result.fields = fields;
  }

  return result;
};

export const parseLogGroupEvents = (logGroup, logStream, logEvents) => {
  const version = parseLambdaVersion(logStream);
  const functionName = parseFunctionName(logGroup);

  const logs = [];

  for (const logEvent of logEvents) {
    try {
      const log = parseLogMessage(logEvent);

      if (log) {
        log.id = logEvent.id;
        log.type = 'cloudwatch';
        log.logStream = logStream;
        log.logGroup = logGroup;
        log.functionName = functionName;
        log.lambdaVersion = version;

        logs.push(log);
      }
    } catch (err) {
      console.error(err.message);
    }
  }

  return logs;
};

export const extractLogs = (eventRecords) => {
  const records = extractRecords(eventRecords);
  return records.reduce((acc, { logGroup, logStream, logEvents }) => {
    const logs = parseLogGroupEvents(logGroup, logStream, logEvents);
    acc.push(...logs);

    return acc;
  }, []);
};
