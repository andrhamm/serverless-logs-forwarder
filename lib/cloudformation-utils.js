import PQueue from 'p-queue';

import { cloudformation } from '../lib/aws-clients';

export const getStackNameForLogGroup = async (logGroupName) => {
  let data;
  try {
    data = await cloudformation.describeStackResources({
      PhysicalResourceId: logGroupName,
    }).promise();
  } catch (e) {
    if (e.message === `Stack for ${logGroupName} does not exist`) {
      return null;
    }

    throw e;
  }

  const resource = data.StackResources.find(r => r.PhysicalResourceId === logGroupName);
  return resource && resource.StackName ? resource.StackName : null;
};

export const getLogGroupsForStack = async (stack, { logGroupNamePrefix }) => {
  const params = {
    StackName: stack.StackName,
    // NextToken: 'STRING_VALUE',
  };

  const allLogGroups = [];

  let data;
  do {
    // eslint-disable-next-line no-await-in-loop
    data = await cloudformation.listStackResources(params).promise();

    const logGroups = data.StackResourceSummaries.filter(r => r.ResourceType === 'AWS::Logs::LogGroup' && r.PhysicalResourceId.startsWith(logGroupNamePrefix));

    allLogGroups.push(...logGroups);

    params.NextToken = data.NextToken;
  } while (params.NextToken);

  return allLogGroups.map(r => r.PhysicalResourceId);
};

export const STACK_STATUS_FILTER = [
  'CREATE_IN_PROGRESS',
  'CREATE_FAILED',
  'CREATE_COMPLETE',
  'ROLLBACK_IN_PROGRESS',
  'ROLLBACK_FAILED',
  'ROLLBACK_COMPLETE',
  'DELETE_IN_PROGRESS',
  'DELETE_FAILED',
  // 'DELETE_COMPLETE',
  'UPDATE_IN_PROGRESS',
  'UPDATE_COMPLETE_CLEANUP_IN_PROGRESS',
  'UPDATE_COMPLETE',
  'UPDATE_ROLLBACK_IN_PROGRESS',
  'UPDATE_ROLLBACK_FAILED',
  'UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS',
  'UPDATE_ROLLBACK_COMPLETE',
  'REVIEW_IN_PROGRESS',
];

export const getLogGroupStackMapping = async (opts) => {
  const params = {
    // NextToken: 'STRING_VALUE',
    StackStatusFilter: STACK_STATUS_FILTER,
  };

  const queue = new PQueue({ concurrency: 4 });

  const mapping = {};

  let data;
  do {
    // eslint-disable-next-line no-await-in-loop
    data = await cloudformation.listStacks(params).promise();

    data.StackSummaries.forEach((stack) => {
      queue.add(() => exports.getLogGroupsForStack(stack, {
        logGroupNamePrefix: opts.logGroupNamePrefix,
      })).then((logGroups) => {
        logGroups.forEach((logGroup) => {
          mapping[logGroup] = stack.StackName;
        });
      });
    });

    params.NextToken = data.NextToken;
  } while (params.NextToken);

  await queue.onIdle();

  return mapping;
};
