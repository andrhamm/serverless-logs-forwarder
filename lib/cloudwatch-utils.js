import { cloudWatchLogs } from './aws-clients';

export const subscribe = ({ logGroupName, roleArn, destinationArn }) =>
  cloudWatchLogs.putSubscriptionFilter({
    destinationArn,
    roleArn,
    logGroupName,
    filterName: 'ship-logs',
    filterPattern: '[timestamp=*Z, request_id="*-*", event]',
  }).promise();

export const getLogGroup = async (logGroupName) => {
  const params = {
    logGroupNamePrefix: logGroupName,
    // nextToken: 'STRING_VALUE',
  };

  let data;
  do {
    // eslint-disable-next-line no-await-in-loop
    data = await cloudWatchLogs.describeLogGroups(params).promise();

    const logGroup = data.logGroups.find(result => result.logGroupName === logGroupName);

    if (logGroup) return logGroup;

    params.nextToken = data.nextToken;
  } while (params.nextToken);

  return {};
};

export const logGroupGetTag = async (logGroupName, tagName) => {
  const tagsResp = await cloudWatchLogs.listTagsLogGroup({ logGroupName }).promise();
  const tags = tagsResp.tags || {};

  return tags[tagName];
};

export const setExpiry = ({ logGroupName, retentionInDays }) =>
  cloudWatchLogs.putRetentionPolicy({
    logGroupName,
    retentionInDays,
  }).promise();

export const getExpiry = async (logGroupName) => {
  const logGroup = await exports.getLogGroup(logGroupName);
  return logGroup.retentionInDays;
};
