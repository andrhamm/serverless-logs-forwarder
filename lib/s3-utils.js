import { tryParseJson } from './common';

import { s3 } from '../lib/aws-clients';

export const tryGetObjectJSON = async (params) => {
  let mappingObject;
  try {
    mappingObject = await s3.getObject(params).promise();
  } catch (e) {
    if (e.code !== 'NoSuchKey') {
      throw e;
    }
  }

  let mapping = {};
  if (mappingObject && mappingObject.Body) {
    const decoded = tryParseJson(mappingObject.Body);

    if (decoded) mapping = decoded;
  }

  return mapping;
};

export const putObjectJSON = (params, mapping) => s3.putObject({
  Body: JSON.stringify(mapping),
  ContentType: 'application/json',
  ...params,
}).promise();
