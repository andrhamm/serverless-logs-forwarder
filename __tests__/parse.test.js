import * as parse from '../lib/parse';
import { kinesisifyRecord, mockKinesisCloudWatchLogsEventRecords } from './__helpers__/kinesis';

afterEach(() => { jest.resetAllMocks(); });

describe('parse', () => {
  describe('parsePayload', () => {
    it('decodes and unzips record payload', () => {
      const expected = mockKinesisCloudWatchLogsEventRecords[0];

      const record = kinesisifyRecord(expected);

      const result = parse.parsePayload(record);

      expect(result).toEqual(expected);
    });
  });

  describe('extractRecords', () => {
    it('returns filtered and decoded records', () => {
      const eventRecords = mockKinesisCloudWatchLogsEventRecords.map(kinesisifyRecord);

      const result = parse.extractRecords(eventRecords);

      expect(result).toMatchSnapshot();
    });
  });

  describe('extractLogs', () => {
    it('returns parsed log objects', () => {
      const records = mockKinesisCloudWatchLogsEventRecords;

      const eventRecords = records.map(kinesisifyRecord);

      const result = parse.extractLogs(eventRecords);

      expect(result).toMatchSnapshot();
    });
  });

  describe('parseFunctionName', () => {
    it('returns lambda name from log group name', () => {
      expect(parse.parseFunctionName('/aws/lambda/my-stack-my_function')).toMatchSnapshot();
    });
  });

  describe('parseLambdaVersion', () => {
    it('returns lambda version from log stream name', () => {
      const logStream = '2018/11/05/[$LATEST]a3a206b7b35f414f9f567242385ede8a';
      const result = parse.parseLambdaVersion(logStream);

      expect(result).toEqual('$LATEST');
    });
  });

  describe('parseLogMessage', () => {
    it('returns null for entry missing id', () => {
      expect(parse.parseLogMessage({
        extractedFields: {
          event: 'my event\n',
        },
      })).toBeNull();
    });

    it('returns null for START entry', () => {
      expect(parse.parseLogMessage({
        id: 1,
        extractedFields: {
          event: 'START RequestId: 67c005bb-641f-11e6-b35d-6b6c651a2f01 Version: 31\n',
        },
      })).toBeNull();
    });

    it('returns null for END entry', () => {
      expect(parse.parseLogMessage({
        id: 1,
        extractedFields: {
          event: 'END RequestId: 5e665f81-641f-11e6-ab0f-b1affae60d28\n',
        },
      })).toBeNull();
    });

    it('returns null for REPORT entry', () => {
      expect(parse.parseLogMessage({
        id: 1,
        extractedFields: {
          event: 'REPORT RequestId: 5e665f81-641f-11e6-ab0f-b1affae60d28\tDuration: 1095.52 ms\tBilled Duration: 1100 ms \tMemory Size: 128 MB\tMax Memory Used: 32 MB\t\n',
        },
      })).toBeNull();
    });

    it('parses normal log message', () => {
      const logEvent = {
        id: 1,
        extractedFields: {
          request_id: 4467343,
          event: 'my event',
          timestamp: 1541438252943,
        },
      };

      const result = parse.parseLogMessage(logEvent);

      expect(result).toEqual({
        level: 'info',
        message: 'my event',
        request_id: 4467343,
        '@timestamp': new Date(1541438252943),
      });
    });

    it('parses json object as fields', () => {
      const jsonLogBody = {
        level: 'warn',
        message: 'my event',
        customField: 'something else',
      };

      const logEvent = {
        id: 1,
        extractedFields: {
          request_id: 4467343,
          event: JSON.stringify(jsonLogBody),
          timestamp: 1541438252943,
        },
      };

      const result = parse.parseLogMessage(logEvent);

      expect(result).toEqual({
        level: 'warn',
        message: 'my event',
        request_id: 4467343,
        fields: {
          customField: 'something else',
        },
        '@timestamp': new Date(1541438252943),
      });
    });
  });
});
