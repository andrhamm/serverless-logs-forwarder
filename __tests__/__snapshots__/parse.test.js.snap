// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse extractLogs returns parsed log objects 1`] = `
Array [
  Object {
    "@timestamp": 2015-05-28T15:27:35.000Z,
    "functionName": "my-stack-my_function",
    "id": "1",
    "lambdaVersion": "$LATEST",
    "level": "info",
    "logGroup": "/aws/lambda/my-stack-my_function",
    "logStream": "2018/11/05/[$LATEST]a3a206b7b35f414f9f567242385ede8a",
    "message": "message 1",
    "request_id": 1,
    "type": "cloudwatch",
  },
  Object {
    "@timestamp": 2015-05-28T15:27:35.000Z,
    "functionName": "my-stack-my_function",
    "id": "2",
    "lambdaVersion": "$LATEST",
    "level": "info",
    "logGroup": "/aws/lambda/my-stack-my_function",
    "logStream": "2018/11/05/[$LATEST]a3a206b7b35f414f9f567242385ede8a",
    "message": "message 2",
    "request_id": 2,
    "type": "cloudwatch",
  },
  Object {
    "@timestamp": 2015-05-28T15:27:35.000Z,
    "fields": Object {
      "customField": "custom field value",
    },
    "functionName": "my-stack-my_function",
    "id": "3",
    "lambdaVersion": "$LATEST",
    "level": "info",
    "logGroup": "/aws/lambda/my-stack-my_function",
    "logStream": "2018/11/05/[$LATEST]a3a206b7b35f414f9f567242385ede8b",
    "message": null,
    "request_id": 3,
    "type": "cloudwatch",
  },
  Object {
    "@timestamp": 2015-05-28T15:27:35.000Z,
    "functionName": "my-stack-my_function",
    "id": "4",
    "lambdaVersion": "$LATEST",
    "level": "info",
    "logGroup": "/aws/lambda/my-stack-my_function",
    "logStream": "2018/11/05/[$LATEST]a3a206b7b35f414f9f567242385ede8b",
    "message": "message 4",
    "request_id": 4,
    "type": "cloudwatch",
  },
]
`;

exports[`parse extractRecords returns filtered and decoded records 1`] = `
Array [
  Object {
    "logEvents": Array [
      Object {
        "extractedFields": Object {
          "event": "message 1",
          "request_id": 1,
          "timestamp": 1432826855000,
        },
        "id": "1",
      },
      Object {
        "extractedFields": Object {
          "event": "message 2",
          "request_id": 2,
          "timestamp": 1432826855000,
        },
        "id": "2",
      },
    ],
    "logGroup": "/aws/lambda/my-stack-my_function",
    "logStream": "2018/11/05/[$LATEST]a3a206b7b35f414f9f567242385ede8a",
    "messageType": "DATA_MESSAGE",
  },
  Object {
    "logEvents": Array [
      Object {
        "extractedFields": Object {
          "event": "{\\"customField\\":\\"custom field value\\"}",
          "request_id": 3,
          "timestamp": 1432826855000,
        },
        "id": "3",
      },
      Object {
        "extractedFields": Object {
          "event": "message 4",
          "request_id": 4,
          "timestamp": 1432826855000,
        },
        "id": "4",
      },
    ],
    "logGroup": "/aws/lambda/my-stack-my_function",
    "logStream": "2018/11/05/[$LATEST]a3a206b7b35f414f9f567242385ede8b",
    "messageType": "DATA_MESSAGE",
  },
]
`;

exports[`parse parseFunctionName returns lambda name from log group name 1`] = `"my-stack-my_function"`;
